File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def maxDepth(self, root: Optional[TreeNode]) -> int:
    def in_order_dfs(root, depth):
      if not root:
        return depth

      left_depth = in_order_dfs(root.left, depth + 1)
      right_depth = in_order_dfs(root.right, depth + 1)

      return max(left_depth, right_depth)

    return in_order_dfs(root, 0)

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
  function dfs(node, depth) {
    if (node === null) return depth;

    const left = dfs(node.left, depth + 1);
    const right = dfs(node.right, depth + 1);

    return Math.max(left, right);
  }

  return dfs(root, 0);
};
